src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -1178,6 +1181,66 @@ DeleteShardPlacementRow(uint64 placementId)
 }
 
 
+/*
+ * UpdatePartitionShardPlacementStates gets a shard placement which is asserted to belong
+ * to partitioned table. The function goes over the corresponding placements of its
+ * partitions, and sets their state to the input shardState.
+ */
+void
+UpdatePartitionShardPlacementStates(ShardPlacement *parentShardPlacement, char shardState)
+{
+	ShardInterval *parentShardInterval =
+		LoadShardInterval(parentShardPlacement->shardId);
+	Oid partitionedTableOid = parentShardInterval->relationId;
+
+	/* this function should only be called for partitioned tables */
+	Assert(PartitionedTable(partitionedTableOid));
+
+	ListCell *partitionOidCell = NULL;
+	List *partitionList = PartitionList(partitionedTableOid);
+	foreach(partitionOidCell, partitionList)
+	{
+		Oid partitionOid = lfirst_oid(partitionOidCell);
+		uint64 partitionShardId =
+			ColocatedShardIdInRelation(partitionOid, parentShardInterval->shardIndex);
+		ShardInterval *partitionShardInterval = LoadShardInterval(partitionShardId);
+
+		ShardPlacement *partitionPlacement =
+			ShardPlacementOnNode(partitionShardInterval, parentShardPlacement->nodeId);
+
+		/* the partition should have a placement on the */
+		Assert(partitionPlacement != NULL);
+
+		UpdateShardPlacementState(partitionPlacement->placementId, shardState);
+	}
+}
+
+
+/*
+ * ShardPlacementOnNode gets a shardInterval and a nodeId, returns a placement
+ * of the shard on the given node. If no such placement exists, the function
+ * return NULL.
+ */
+static ShardPlacement *
+ShardPlacementOnNode(ShardInterval *shardInterval, int nodeId)
+{
+	List *placementList = ShardPlacementList(shardInterval->shardId);
+	ListCell *placementCell = NULL;
+
+	foreach(placementCell, placementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+
+		if (placement->nodeId == nodeId)
```

Comment:
Would be nicer to use groupId here. Placements do not have a node ID  in the metadata. It just so happens that we applied the same groupId->nodeId mapping to both of these placements.

Reply:
Would perhaps be nice to have a GroupShardPlacementList instead of ShardPlacementList.